server/graphql/common/uploaded-file.ts

```diff
@@ -0,0 +1,110 @@
+import { Request } from 'express';
+
+import { Collective, Expense, ExpenseAttachedFile, ExpenseItem, UploadedFile } from '../../models';
+import { ExpenseStatus } from '../../models/Expense';
+import { idDecode, IDENTIFIER_TYPES } from '../v2/identifiers';
+
+import { canSeeExpenseAttachments, canSeeExpenseDraftPrivateDetails } from './expenses';
+
+export async function hasProtectedUrlPermission(req: Request, url: string) {
+  const requestedUrl = new URL(url);
+  const encodedExpenseId = requestedUrl.searchParams.get('expenseId');
+  const draftKey = requestedUrl.searchParams.get('draftKey');
+  requestedUrl.hash = '';
+  requestedUrl.search = '';
+
+  const protectedUrl = requestedUrl.toString();
+
+  const uploadedFile = await UploadedFile.getFromProtectedURL(protectedUrl);
+
+  if (!uploadedFile) {
+    return false;
+  }
+
+  let expenseId: number;
+  if (encodedExpenseId) {
+    expenseId = idDecode(encodedExpenseId, IDENTIFIER_TYPES.EXPENSE);
+  }
+
+  return hasUploadedFilePermission(req, uploadedFile, {
+    expenseId,
+    draftKey,
+  });
+}
+
+export async function hasUploadedFilePermission(
+  req: Request,
+  uploadedFile: UploadedFile,
+  options?: { expenseId: number; draftKey?: string },
+): Promise<boolean> {
+  const actualUrl = uploadedFile.getDataValue('url');
+
+  const expense = options.expenseId && (await Expense.findByPk(options.expenseId));
+
+  if (expense && expense.status === ExpenseStatus.DRAFT) {
+    const itemMatches = ((expense.data?.items as { url?: string }[]) || []).some(
+      item => item.url && item.url === actualUrl,
+    );
+    const attachedFileMatches = ((expense.data?.attachedFiles as { url?: string }[]) || []).some(
+      item => item.url && item.url === actualUrl,
+    );
+
+    const fileIsInDraft = itemMatches || attachedFileMatches;
+
+    if (fileIsInDraft && (await canSeeExpenseDraftPrivateDetails(req, expense))) {
+      return true;
+    }
+
+    if (fileIsInDraft && options.draftKey && expense.data.draftKey === options.draftKey) {
+      return true;
+    }
+  }
+
+  if (!req?.remoteUser) {
+    return false;
+  }
+
+  switch (uploadedFile.kind) {
+    case 'EXPENSE_ITEM': {
+      const expenseItem = await ExpenseItem.findOne({
+        where: {
+          url: actualUrl,
+        },
+        include: { model: Expense, include: [{ association: 'fromCollective' }] },
+      });
+
+      const expense = expenseItem?.Expense;
+
+      if (!expense) {
+        return req.remoteUser?.id === uploadedFile.CreatedByUserId;
+      }
```

Comment:
How can an expense item exist without an expense? Did you mean to write:

```suggestion
      const expense = expenseItem?.Expense;

      if (!expenseItem) {
        return req.remoteUser?.id === uploadedFile.CreatedByUserId;
      }
```